### PR TITLE
extended the existing-coverage option to support bytecode_index

### PR DESCRIPTION
**Pull request by lucasccordeiro**
Extended the existing-coverage option to support bytecode_index, 
given the current JSON format to specify goals for C and Java programs.
This PR https://github.com/diffblue/test-gen/pull/760 contains the regression 
tests for C and Java programs to check the existing-coverage option

### CODE REVIEW

File: src/goto-instrument/cover.cpp [hunk 1]
```diff
@@ -132,24 +132,33 @@ bool coverage_goalst::get_coverage_goals(
...
-      for(const auto &line_json : goal["lines"].array)
+      // store the source location for each existing goal
+      for(const auto &each_goal : goals_in_json["goals"].array)
       {
-        // get the line of each existing goal
-        irep_idt line=line_json["number"].value;
+        // get and set the bytecode_index
+        irep_idt bytecode_index=
+          each_goal["sourceLocation"]["bytecode_index"].value;
+        source_location.set_java_bytecode_index(bytecode_index);
```
Comment (mgudemann) [inline src/goto-instrument/cover.cpp]: this should be accessed only if the `bytecode_index` field is present which is only true for Java. Best would be an invariant that requires `byecode_index` in Java mode.

Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Now we check whether we have empty fields for `bytecode_index`, `file`, `function` and `line` before accessing the json data structure. In particular, we add as precondition the following statement as suggested by @peterschrammel:

````
        PRECONDITION(
          (!each_goal["goal"].is_null()) ||
          (!each_goal["sourceLocation"]["bytecode_index"].is_null()) ||
          (!each_goal["sourceLocation"]["file"].is_null() &&
           !each_goal["sourceLocation"]["function"].is_null() &&
		   !each_goal["sourceLocation"]["line"].is_null()));
````

Let me know if you have further comments.

**Comment by lucasccordeiro** [comment]
@mgudemann and @peterschrammel: I have addressed all your comments. Let me know if you have additional suggestions.

File: src/goto-instrument/cover.cpp [hunk 2]
```diff
@@ -132,50 +133,106 @@ bool coverage_goalst::get_coverage_goals(
...
+{
+  for(auto const &existing_loc : existing_goals)
+  {
+    if(!existing_loc.second)
+    {
+      msg.warning() << "Warning: existing goal in file "
+      << existing_loc.first.get_file()
+      << " line " << existing_loc.first.get_line()
+      << " function " << existing_loc.first.get_function()
+      << " is uncovered" << messaget::eom;
```
Comment (mgudemann) [inline src/goto-instrument/cover.cpp]: I think this should be reformatted to have the `<<` aligned

File: src/goto-instrument/cover.cpp [hunk 3]
```diff
@@ -1441,14 +1498,20 @@ bool instrument_cover_goals(
   }
 
   // check existing test goals
-  coverage_goalst existing_goals;
+  static coverage_goalst existing_goals;
```
Comment (mgudemann) [inline src/goto-instrument/cover.cpp]: why is this static

File: src/goto-instrument/cover.h [hunk 4]
```diff
@@ -23,12 +23,14 @@ class coverage_goalst
   static bool get_coverage_goals(
     const std::string &coverage,
     message_handlert &message_handler,
-    coverage_goalst &goals);
+    coverage_goalst &goals,
+    irep_idt &mode);
   void add_goal(source_locationt goal);
-  bool is_existing_goal(source_locationt source_loc) const;
+  bool is_existing_goal(source_locationt source_loc);
+  void check_uncovered_goals(messaget msg);
```
Comment (mgudemann) [inline src/goto-instrument/cover.h]: should be `messaget &`

**Comment by lucasccordeiro** [comment]
@peterschrammel: do you still have comments related to this PR?

File: src/goto-instrument/cover.cpp [hunk 5]
```diff
@@ -108,7 +108,8 @@ class basic_blockst
 bool coverage_goalst::get_coverage_goals(
   const std::string &coverage_file,
   message_handlert &message_handler,
-  coverage_goalst &goals)
+  coverage_goalst &goals,
+  irep_idt &mode)
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: `const`

File: src/goto-instrument/cover.cpp [hunk 6]
```diff
@@ -132,50 +133,113 @@ bool coverage_goalst::get_coverage_goals(
...
+        PRECONDITION(
+          (!each_goal["goal"].is_null()) ||
+          (!each_goal["sourceLocation"]["bytecode_index"].is_null()) ||
+          (!each_goal["sourceLocation"]["file"].is_null() &&
+           !each_goal["sourceLocation"]["function"].is_null() &&
+           !each_goal["sourceLocation"]["line"].is_null()));
+
+        // check whether bytecode_index is provided for Java programs
+        if(mode==ID_java)
+          if(each_goal["sourceLocation"]["bytecode_index"].is_null())
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: use `&&`

File: src/goto-instrument/cover.cpp [hunk 7]
```diff
@@ -132,50 +133,113 @@ bool coverage_goalst::get_coverage_goals(
...
+/// check whether we have an existing goal that is uncovered
+/// \param msg: message to be printed about the uncovered goal
+void coverage_goalst::check_uncovered_goals(messaget &msg)
+{
+  for(auto const &existing_loc : existing_goals)
+  {
+    if(!existing_loc.second)
+    {
+      msg.warning()
+      << "Warning: existing goal in file "
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: indentation

File: src/goto-instrument/cover.cpp [hunk 8]
```diff
@@ -1444,11 +1508,17 @@ bool instrument_cover_goals(
   coverage_goalst existing_goals;
   if(cmdline.isset("existing-coverage"))
   {
-    msg.status() << "Check existing coverage goals" << messaget::eom;
+    // get the mode to ensure invariants
+    // (e.g., bytecode_index for Java programs)
+    namespacet ns(symbol_table);
+    irep_idt mode=ns.lookup(goto_functions.entry_point()).mode;
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: use a const ref

File: src/goto-instrument/cover.h [hunk 9]
```diff
@@ -23,12 +23,14 @@ class coverage_goalst
   static bool get_coverage_goals(
     const std::string &coverage,
     message_handlert &message_handler,
-    coverage_goalst &goals);
+    coverage_goalst &goals,
+    irep_idt &mode);
```
Comment (peterschrammel) [inline src/goto-instrument/cover.h]: `const`

File: src/goto-instrument/cover.cpp [hunk 10]
```diff
@@ -166,25 +166,47 @@ bool coverage_goalst::get_coverage_goals(
...
 }
 
 /// compare the value of the current goal to the existing ones
 /// \param source_loc: source location of the current goal
 /// \return true : if the current goal exists false : otherwise
-bool coverage_goalst::is_existing_goal(source_locationt source_loc) const
+bool coverage_goalst::is_existing_goal(source_locationt source_loc)
 {
-  for(const auto &existing_loc : existing_goals)
+  for(auto const &existing_loc : existing_goals)
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: `const auto &`

File: src/goto-instrument/cover.cpp [hunk 11]
```diff
@@ -142,9 +142,9 @@ bool coverage_goalst::get_coverage_goals(
       for(const auto &each_goal : goals_in_json["goals"].array)
       {
         // get and set the bytecode_index
-        irep_idt bytecode_index=
+        irep_idt bytecodeIndex=
```
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: no camelCase, use underscore

**Comment by peterschrammel** [comment]
Please reasonably squash the commits

File: src/goto-instrument/cover.cpp (hunk 5 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

File: src/goto-instrument/cover.cpp (hunk 6 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

File: src/goto-instrument/cover.cpp (hunk 7 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

File: src/goto-instrument/cover.cpp (hunk 8 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

File: src/goto-instrument/cover.h (hunk 9 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.h]: Implemented as suggested.

File: src/goto-instrument/cover.cpp (hunk 10 above)
Comment (lucasccordeiro) [inline src/goto-instrument/cover.cpp]: Implemented as suggested.

**Pull request merged**